components/tabs/ios/Tabs/Tabs.swift

```diff
@@ -6,28 +6,27 @@ import Foundation
 
 open class TabsStorage {
     private var store: TabsStore
-    private let queue = DispatchQueue(label: "com.mozilla.tabs-storage")
 
     public init(databasePath: String) {
         store = TabsStore(path: databasePath)
     }
 
     /// Get all tabs by client.
-    open func getAll() -> [ClientRemoteTabs] {
-        return queue.sync {
+    open func getAll() async -> [ClientRemoteTabs] {
+        Task {
             self.store.getAll()
         }
```

Comment:
I think this is the main question: can we simply replace `DispatchQueue` with `Task` and be done with it?

In particular: `getAll()` runs a database query that could potentially fetch a lot of data.  Would using a `Task` like this block the main thread or other background tasks while the query is running or would they run in parallel?  Should we be using `Task.detached` instead?

Reply:
Task will never run the block of code on the main thread unless you specifically request it to.
That syntax would look like this:
```
Task { @MainActor in
    // Some code
 }
```
For Task.detached it's generally recommended to avoid using these unless you have a compelling reason to do so. There is a good write up here on what the difference is if you want to know more https://www.hackingwithswift.com/quick-start/concurrency/whats-the-difference-between-a-task-and-a-detached-task

Reply:
Oh, thanks for clarifying that! Just to make super sure—will making a blocking I/O call inside a task (detached or not) prevent all other tasks from running until that call is done?

I wanted to double-check because all the examples of Swift concurrency I've seen so far use `await` inside the operation passed to `Task`. These examples also take care to mention that tasks should always make forward progress.

But our case feels a bit different. In our case, we're calling into foreign Rust code that does synchronous disk (and sometimes network) I/O. The Rust code has no awareness of the Swift task runtime, so it can't suspend itself, and can't be canceled until it's done.

At the same time, we want Swift callers to be able to do other things while the Rust code is blocked on I/O. For a concrete example, let's say that `SearchViewController` calls `TabsStorage.getAll()`. We want the controller to be able to run `TabsStorage.getAll()` concurrently with the network call to query Google for search engine suggestions, and with calls to Places to search history and bookmarks. The thing we want to avoid is forcing the controller to wait for `TabsStorage.getAll()` to run to completion before it can start fetching search engine suggestions, or querying bookmarks and history.

Zooming out a bit, I think what I'm struggling to understand with Swift concurrency is how to adapt existing blocking functions to work with it.

With GCD, we gave `TabsStorage` and `RustPlaces` their own private serial queues—not because we need the sequential execution semantics, but to avoid thread explosion with concurrent queues. IIUC, that meant that all `TabsStorage` operations would run serially, and all `RustPlaces` operations would run serially, but `TabsStorage` operations could (but not necessarily guaranteed to) run concurrently with `RustPlaces` operations, and vice versa. Is that mental model accurate?

With Swift concurrency, I'm wondering if that's still the case—just without the need to create our own dispatch queues—or if the task runtime executes _all_ tasks sequentially, and so a single long-running task will block all other tasks from running until it's done.

(So sorry for that long wall of text! 😅 I haven't really found a good explanation for this, and it would be amazing to have your insights and guidance, when you have some time—and thank you!)